taglocker_test.go

```diff
@@ -0,0 +1,162 @@
+// paths to that tag but also all intermediary prefixes.
+// We will test that we can lock intermediary nodes by
+// randomly choosing from these returned values.
+func genPaths() []string {
+	var paths []string
+
+	/*
+	* The EBNF is:
+	*
+	* tag ::= SYMBOLIC_SEG ( tag_seg )* ( bit_seg )?
+	*
+	* tag_seg ::= '.' SYMBOLIC_SEG
+	*             '[' array_seg ']'
+	*
+	* bit_seg ::= '.' [0-9]+
+	*
+	* array_seg ::= NUMERIC_SEG ( ',' NUMERIC_SEG )*
+	*
+	* SYMBOLIC_SEG ::= [a-zA-Z]([a-zA-Z0-9_]*)
+	*
+	* NUMERIC_SEG ::= [0-9]+
+	*
+	 */
+	genSymbolicSeg := func() string {
+		length := 1 + rand.Intn(10)
+		bs := make([]byte, length)
+		bs[0] = alphaCharset[rand.Intn(len(alphaCharset))]
+		for i := 1; i < len(bs); i++ {
+			bs[i] = alphanumCharset[rand.Intn(len(alphanumCharset))]
+		}
+		return string(bs)
+	}
+
+	genNumericSeg := func() string {
+		return fmt.Sprintf("%d", rand.Intn(10))
+	}
+	genArraySeg := func() string {
+		// TODO: when the parser supports comma-separated multidimensional
+		// arrays, include potentially generating those.
+		return genNumericSeg()
+	}
+	genTagSeg := func() string {
+		if rand.Intn(2) == 0 {
+			return "." + genSymbolicSeg()
+		}
+		return fmt.Sprintf("[%s]", genArraySeg())
+	}
+	// tag ::= SYMBOLIC_SEG ( tag_seg )* ( bit_seg )?
+	var sb strings.Builder
+	sb.WriteString(genSymbolicSeg())
+	for i := 0; i < rand.Intn(5); i++ {
+		//paths = append(paths, sb.String())
+		sb.WriteString(genTagSeg())
+	}
+	paths = append(paths, sb.String())
+	return paths
+}
+
+const serialConcurrency = 1
+const lowConcurrency = 2
+const medConcurrency = 5
+const highConcurrency = 10
+const writePerc = 1
+const heavyWritePerc = 10
+
+func BenchmarkSerialTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, serialConcurrency, writePerc)
+}
+
+func BenchmarkSerialTagLockingWithHeavyWrites(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, serialConcurrency, heavyWritePerc)
+}
+
+func BenchmarkLowConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, lowConcurrency, writePerc)
+}
+
+func BenchmarkMedConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, medConcurrency, writePerc)
+}
+
+func BenchmarkHighConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, highConcurrency, writePerc)
+}
+
+func BenchmarkHighConcurrencyTagLockingWithHeavyWrites(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, highConcurrency, heavyWritePerc)
+}
+
+func benchmarkTagLockLocking(b *testing.B, tests int, concurrency int, writePerc int) {
+	barrier := make(chan bool, concurrency)
+
+	tl := NewTagLocker(&NoOpReadWriter{})
+
+	read := func(name string) {
+		err := tl.ReadTag(name, 42)
+		if err != nil {
+			b.Errorf("%v\n", err)
+		}
+		<-barrier
+	}
+	write := func(name string) {
+		err := tl.WriteTag(name, 42)
+		if err != nil {
+			b.Errorf("%v\n", err)
+		}
+		<-barrier
+	}
+
+	var tags []string
+	for i := 0; i < 10; i++ {
+		tags = append(tags, genPaths()...)
+	}
+
```

Comment:
If the preceding setup isn't fast, you might want to call `b.ResetTimer()` here. (I'm guessing it doesn't matter.)

Reply:
True, but I think having the ResetTimer there is not bad documentation in terms of "the real test begins from here on out", and as you say it might make the benchmark marginally more correct.